packages/hub/indexing/document-context.js

```diff
@@ -107,6 +108,203 @@ module.exports = class DocumentContext {
     return resource;
   }
 
+  // This takes an arbitrary JSON:API document and makes it safe to
+  // show within the given context. If it can't be shown at all
+  // (because the primary resource is not readable), we return
+  // undefined.
+  //
+  // Otherwise, we go through each included resource and field to
+  // ensure there are valid grants for them. We return a new sanitized
+  // document that strips out any included resources or fields that
+  // were lacking grants.
+  async applyReadAuthorization(document, context={}) {
+    // `context` here is not a DocumentContext object but extra
+    // but a POJO of attributes (for example, { id, type, session })
+    if (!document.data) {
+      return;
+    }
+
+    let session = context.session || Session.EVERYONE;
+    let userRealms = await session.realms();
+
+    let authorizedResource;
+    let types = this.schema.types;
+    if (Array.isArray(document.data)) {
+      authorizedResource = document.data.map(resource => {
+        if (resource.id == null || resource.type == null) {
+          return;
+        }
+
+        let type = types.get(resource.type);
+        if (!type) {
+          return;
+        }
+        return type.applyReadAuthorization(resource, userRealms);
+      }).filter(Boolean);
+    } else {
+      if (document.data.id == null || document.data.type == null) {
+        return;
+      }
+
+      let primaryType = types.get(document.data.type);
+      if (!primaryType) {
+        return;
+      }
+      if (document.data.type === 'permissions') {
+        authorizedResource = await this._readAuthorizationForPermissions(document, session, context);
+      } else {
+        authorizedResource = primaryType.applyReadAuthorization(document.data, userRealms);
+        if (!authorizedResource) {
+          return;
+        }
+      }
+    }
+
+    let output = document;
+
+    if (document.data !== authorizedResource) {
+      output = {
+        data: authorizedResource
+      };
+      if (document.meta) {
+        output.meta = document.meta;
+      }
+    }
+
+    if (document.included) {
+      let safeIncluded = this._readAuthIncluded(document.included, userRealms);
+      if (safeIncluded === document.included) {
```

Comment:
@balinterdi did you mean a deep equal here? because this is not testing for the object equality

Reply:
(and also since this is `included` you might actually want to test for `Set` equality, as opposed to `Array` equality.

Reply:
See Answer 1.